app/models/spree/order_decorator.rb

```diff
@@ -144,6 +144,8 @@ def add_variant(variant, quantity = 1, max_quantity = nil, currency = nil)
   end
 
   # After changing line items of a completed order
+  # TODO: perhaps this should be triggered from a controller
```

Comment:
I think I had a similar thought when I last looked at this code.

Reply:
I don't like to put it neither in the model as a callback nor the controller. Big classes are unmaintenable. Let's extract this into a new abstraction like a service object.

Reply:
@sauloperez I’m always happy to abstract detail down into a service object. 👍 My comment was just a suggestion that a controller action is perhaps a more appropriate hook.

Reply:
as I hook, yes. I'm just troubled with so many nasty callbacks :see_no_evil:

Reply:
Me too. 😔 Unfortunately we inherit a lot of them from Spree, but there are plenty of our own that we can work on removing! 💪